clients/da-vinci-client/src/main/java/com/linkedin/davinci/consumer/VeniceChangelogConsumerImpl.java

```diff
@@ -620,6 +621,21 @@ protected Collection<PubSubMessage<K, ChangeEvent<V>, VeniceChangeCoordinate>> i
     if (changeCaptureStats != null) {
       changeCaptureStats.recordRecordsConsumed(pubSubMessages.size());
     }
+    if (changelogClientConfig.shouldCompactMessages()) {
+      Map<K, PubSubMessage<K, ChangeEvent<V>, VeniceChangeCoordinate>> tempMap = new LinkedHashMap<>();
+      // The behavior of LinkedHashMap is such that it maintains the order of insertion, but for values which are
+      // replaced,
+      // it's put in at the position of the first insertion. This isn't quite what we want, we want to keep only
+      // a single key (just as a map would), but we want to keep the position of the last insertion as well. So in order
+      // to do that, we remove the entry before inserting it.
+      for (PubSubMessage<K, ChangeEvent<V>, VeniceChangeCoordinate> message: pubSubMessages) {
+        if (tempMap.containsKey(message.getKey())) {
```

Comment:
In your test case, `K` is `String`, which is comparable as expected.
But in this file, key is `byte[]`. byte[] arrays are compared based on reference equality rather than content equality.

Reply:
That's an interesting point. The type of message.getKey will be parameterized by the type given in the client...  So it would imply that that type needs to be something that implements equals(), which probably won't be universally true.  I know in java we have Arrays.equals() for this sort of thing, but then we'd have to peek under the covers a bit.....  I'm not totally sure how to overcome this